File: lin-compiler/lin_compiler/Opcode/BaseOpcode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LIN
{
    class BaseOpcode(byte opcode, string name, params Type[] paramTypes)
    {
        public string Name = name;
        public byte Opcode = opcode;
        public Type[] ParamTypes = paramTypes ?? [];
        public bool IsVarArg = false;

        // Constructor for simple byte-only opcodes
        public BaseOpcode(byte opcode, string name, int byteCount) : this(opcode, name, [.. Enumerable.Repeat(Type.Byte, byteCount)])
        {
        }

        // Constructor with opcode defaulting to 0xFF
        public BaseOpcode(string name, params Type[] paramTypes) : this(byte.MaxValue, name, paramTypes)
        {
        }

        // Constructor with opcode defaulting to 0xFF
        public BaseOpcode(string name, int byteCount) : this(byte.MaxValue, name, [.. Enumerable.Repeat(Type.Byte, byteCount)])
        {
        }

        // Constructor with just byte count, name will be hex representation
        public BaseOpcode(int byteCount) : this(byte.MaxValue, null, [.. Enumerable.Repeat(Type.Byte, byteCount)])
        {
        }

        // Constructor with just param types, name will be hex representation
        public BaseOpcode(params Type[] paramTypes) : this(byte.MaxValue, null, paramTypes)
        {
        }

        public int GetByteCount()
        {
            int count = 0;
            foreach (var type in ParamTypes)
            {
                count += Parameter.CountBytes(type);
            }
            return count;
        }

        public void WriteSource(StringBuilder output, Script script, ScriptEntry scriptEntry)
        {
            string opcodeName = Program.UseHexOpcodes ? $"0x{Opcode:X2}" : (Name ?? $"0x{Opcode:X2}");
            output.Append(opcodeName);
            output.Append('(');
            WriteSourceArgs(output, script, scriptEntry);
            output.AppendLine(")");
        }

        protected virtual void WriteSourceArgs(StringBuilder output, Script script, ScriptEntry scriptEntry)
        {
            byte[] argValues = scriptEntry.Args;
            if (argValues.Length == 0)
            {
                return;
            }

            List<string> args = [];
            int byteIndex = 0;

            foreach (var paramType in paramTypes)
            {
                args.Add(Parameter.FormatValue(paramType, argValues, ref byteIndex));
            }
            output.AppendJoin(", ", args);
        }

        public virtual List<ScriptEntry> ReadSource(string argsString, int lineNum, Script script)
        {
            return [new ScriptEntry()
            {
                Opcode = Opcode,
                Args = ParseOpcodeArgs(argsString, lineNum),
            }];
        }
        
        protected virtual byte[] ParseOpcodeArgs(string argsString, int lineNum)
        {
            argsString = argsString.Trim();

            // Empty args
            if (string.IsNullOrWhiteSpace(argsString))
            {
                return [];
            }

            // Split by comma
            string[] argStrings = argsString.Split(',');
            for (int i = 0; i < argStrings.Length; i++)
            {
                argStrings[i] = argStrings[i].Trim();
            }

            // Convert arguments to bytes based on parameter types
            List<byte> argBytes = [];

            int argIndex = 0;
            foreach (var paramType in ParamTypes)
            {
                if (argIndex >= argStrings.Length)
                {
                    throw new Exception($"[read] error: not enough arguments at line {lineNum + 1}");
                }

                Parameter.ParseValue(paramType, argStrings[argIndex], argBytes, lineNum);
                argIndex++;
            }

            return [.. argBytes];
        }

        // Called during compilation to prepare the entry for writing
        // Override this method for opcodes that need special preparation (e.g., text ID assignment)
        public virtual void PrepareForCompilation(Script script, ScriptEntry entry)
        {
            // Default: no special preparation needed
        }

    };
}

File: lin-compiler/src/opcodes/baseOpcode.ts
import { options, toHexOpcode } from "../options.js";
import type { SourceBuilder } from "../output.js";
import { type ByteCursor, countBytes, formatValue, ParamType, parseValue } from "../parameter.js";
import type { Script, ScriptEntry } from "../script.js";

/** Shorthand for an opcode that takes `count` plain bytes. */
export function bytes(count: number): ParamType[] {
  return new Array(count).fill(ParamType.Byte);
}

export class BaseOpcode {
  /** Set when the opcode is registered in the dictionary. */
  opcode: number;
  readonly name: string | null;
  readonly paramTypes: ParamType[];
  isVarArg = false;

  constructor(name: string | null, paramTypes: ParamType[] | number = [], opcode = 0xff) {
    this.opcode = opcode;
    this.name = name;
    this.paramTypes = typeof paramTypes === "number" ? bytes(paramTypes) : paramTypes;
  }

  getByteCount(): number {
    return this.paramTypes.reduce((total, type) => total + countBytes(type), 0);
  }

  get displayName(): string {
    return this.name ?? toHexOpcode(this.opcode);
  }

  writeSource(output: SourceBuilder, script: Script, scriptEntry: ScriptEntry): void {
    output.append(options.useHexOpcodes ? toHexOpcode(this.opcode) : this.displayName);
    output.append("(");
    this.writeSourceArgs(output, script, scriptEntry);
    output.appendLine(")");
  }

  writeSourceArgs(output: SourceBuilder, _script: Script, scriptEntry: ScriptEntry): void {
    const argValues = scriptEntry.args;
    if (argValues.length === 0) {
      return;
    }

    const args: string[] = [];
    const cursor: ByteCursor = { index: 0 };
    const values = Uint8Array.from(argValues);

    for (const paramType of this.paramTypes) {
      args.push(formatValue(paramType, values, cursor));
    }
    output.appendJoin(", ", args);
  }

  readSource(argsString: string, lineNum: number, _script: Script): ScriptEntry[] {
    return [
      {
        opcode: this.opcode,
        args: this.parseOpcodeArgs(argsString, lineNum),
      },
    ];
  }

  protected parseOpcodeArgs(argsString: string, lineNum: number): number[] {
    const trimmed = argsString.trim();

    // Empty args
    if (trimmed.length === 0) {
      return [];
    }

    const argStrings = trimmed.split(",").map((arg) => arg.trim());

    // Convert arguments to bytes based on parameter types
    const argBytes: number[] = [];

    let argIndex = 0;
    for (const paramType of this.paramTypes) {
      if (argIndex >= argStrings.length) {
        throw new Error(`[read] error: not enough arguments at line ${lineNum + 1}`);
      }

      parseValue(paramType, argStrings[argIndex], argBytes, lineNum);
      argIndex++;
    }

    return argBytes;
  }

  /**
   * Called during compilation to prepare the entry for writing.
   * Override for opcodes that need special preparation (e.g. text ID assignment).
   */
  prepareForCompilation(_script: Script, _entry: ScriptEntry): void {
    // Default: no special preparation needed
  }
}
